Sources/SwiftDocC/Model/Rendering/HTML/HTMLRenderer.swift

```diff
@@ -0,0 +1,291 @@
+                        value: convert(subheading)
+                    )},
+                    uniquingKeysWith: { _, new in new }
+                ))
+            } else {
+                // There are multiple subheadings, but the're all the same
+                subheadings = .single(.symbol(convert(primarySubheading ?? [])))
+            }
+        } else {
+            subheadings = .single(.conceptual(node.name.plainText))
+        }
+        
+        return .init(
+            path: Self.filePath(for: node.reference),
+            names: names,
+            subheadings: subheadings,
+            abstract: (node.semantic as? (any Abstracted))?.abstract
+        )
+    }
+    
+    func pathForSymbolID(_ usr: String) -> URL? {
+        context.localOrExternalReference(symbolID: usr).map {
+            Self.filePath(for: $0)
+        }
+    }
+    
+    func assetNamed(_ assetName: String) -> LinkedAsset? {
+        guard let asset = context.resolveAsset(named: assetName, in: reference) else {
+            // The context
+            return nil
+        }
+        
+        var files = [LinkedAsset.ColorStyle: [Int: URL]]()
+        for (traits, url) in asset.variants {
+            let scale = (traits.displayScale ?? .standard).scaleFactor
+            
+            files[traits.userInterfaceStyle == .dark ? .dark : .light, default: [:]][scale] = url
+        }
+        
+        return .init(files: files)
+    }
+    
+    func fallbackLinkText(linkString: String) -> String {
+        // For unresolved links, especially to symbols, prefer to display only the the last link component without its disambiguation
+        PathHierarchy.PathParser.parse(path: linkString).components.last.map { String($0.name) } ?? linkString
+    }
+    
+    static func filePath(for reference: ResolvedTopicReference) -> URL {
+        reference.url.withoutHostAndPortAndScheme().appendingPathComponent("index.html")
+    }
+}
+
+/// A type that renders documentation pages into semantic HTML elements.
+struct HTMLRenderer {
+    let reference: ResolvedTopicReference
+    let context: DocumentationContext
+    let goal: RenderGoal
+    
+    private let renderer: MarkdownRenderer<ContextLinkProvider>
+    
+    init(reference: ResolvedTopicReference, context: DocumentationContext, goal: RenderGoal) {
+        self.reference = reference
+        self.context = context
+        self.goal = goal
+        self.renderer = MarkdownRenderer(
+            path: ContextLinkProvider.filePath(for: reference),
+            goal: goal,
+            linkProvider: ContextLinkProvider(reference: reference, context: context)
+        )
+    }
+    
+    /// Information about a rendered page
+    struct RenderedPageInfo {
+        /// The HTML content of the page as an XMLNode hierarchy.
+        ///
+        /// The string representation of those node hierarchy is intended to be inserted _somewhere_ inside the `<body>` HTML element.
```

Comment:
```suggestion
        /// The string representation of this node hierarchy is intended to be inserted _somewhere_ inside the `<body>` HTML element.
```

Also, should we split this doc comment up into an abstract and a note, or overview?